mill-universal-packager/src/io/github/hoangmaihuy/mill/packager/universal/Archiver.scala

```diff
@@ -0,0 +1,169 @@
+  ): Unit = {
+    Using.resource(new TarArchiveOutputStream(compress(new BufferedOutputStream(new FileOutputStream(out.toIO))))) {
+      tos =>
+        mappings.foreach { case (from, to) =>
+          addPathToTar(from, to, tos, wrappedIn)
+        }
+    }
+  }
+
+  // recursively add a Path which can be a file or a directory to the given TarArchiveOutputStream.
+  // if the path is a single file, just add it to the output stream,
+  // if the path is a directory, add all of the contents under the path to the output stream
+  private def addPathsToTarball(path: Path, outputStream: OutputStream, wrappedIn: Option[SubPath]): Unit = {
+    Using.resource(new TarArchiveOutputStream(outputStream)) { tarArchive =>
+      path match {
+        case _ if os.isDir(path) =>
+          os.walk(path).foreach { p =>
+            addPathToTar(
+              path = p,
+              target = p.subRelativeTo(path),
+              tarArchive = tarArchive,
+              wrappedIn = wrappedIn
+            )
+          }
+        case _ if os.isFile(path) =>
+          addPathToTar(
+            path = path,
+            target = path.subRelativeTo(path / os.up),
+            tarArchive = tarArchive,
+            wrappedIn = wrappedIn
+          )
+        // this should not happen
+        case _ =>
+      }
+    }
+  }
+
+  // add single Path which can be a file or a directory to the given TarArchiveOutputStream, not recursively
+  private def addPathToTar(
+    path: Path,
+    target: SubPath,
+    tarArchive: TarArchiveOutputStream,
+    wrappedIn: Option[SubPath]
+  ): Unit = {
+    val file: File = path.toIO
+    // Create entry name relative to parent file path
+    val entryPath = wrappedIn.map(_ / target).getOrElse(target)
+    val tarEntry = new TarArchiveEntry(file, entryPath.toString)
+    SOURCE_DATE_EPOCH.foreach { d =>
+      // set the last modified time as needed
+      tarEntry.setLastModifiedTime(FileTime.from(d, TimeUnit.SECONDS))
+    }
+    // set the executable permission for files under the /bin directory
+    if ((path / os.up).baseName == "bin") {
+      tarEntry.setMode(EXEXCUTABLE_FILE_PERMSET.value)
+    }
+    // add tar ArchiveEntry
+    tarArchive.putArchiveEntry(tarEntry)
+    if (os.isFile(path)) {
+      Using.resource(new BufferedInputStream(new FileInputStream(file))) { inputStream =>
+        // Write file content to archive
+        copyInputStreamToOutputStream(inputStream, tarArchive)
+      }
+    }
+    tarArchive.closeArchiveEntry()
+  }
+
+  /** Reads all bytes from an input stream and writes them to an output stream.
+    *
+    * TODO: use in.transferTo(out) once we drop JAVA 8 support.
+    */
+  private def copyInputStreamToOutputStream(in: InputStream, out: OutputStream): Long = {
+    var nread = 0L
+    val buf = new Array[Byte](BUFFER_SIZE)
+    var n = 0
+    while ({ n = in.read(buf); n } > 0) {
+      out.write(buf, 0, n)
+      nread += n
+    }
+    nread
+  }
```

Comment:
or use the Apache Commons IOUtils.copy() instead.
or shall we just not care the JAVA 8 now?

Reply:
I think it's better to declare our own method, sbt did that. Here is `transfer` implementation from sbt that I used in other projects
```
  private val BufferSize = 8192

  private def transfer(in: InputStream, out: OutputStream, close: Boolean) = {
    try {
      val buffer = new Array[Byte](BufferSize)

      @tailrec def read(): Unit = {
        val byteCount = in.read(buffer)
        if (byteCount >= 0) {
          out.write(
            buffer,
            0,
            byteCount
          )
          read()
        }
      }

      read()
    } finally {
      if (close) in.close()
    }
  }
```